src/python/pants/java/nailgun_client.py

```diff
@@ -137,6 +143,9 @@ def _process_session(self):
             self._remote_pgrp_callback(self.remote_pgrp)
         elif chunk_type == ChunkType.START_READING_INPUT:
           self._maybe_start_input_writer()
+        elif chunk_type == ChunkType.CLIENT_BLOCKED:
+          if self._fallback_on_client_blocked:
+            raise RemoteClientFallback()
```

Comment:
Caveat: I'm looking at this code without a lot of prior context, so feel free to follow your best judgement over my not-very-informed opinion.

To me this path doesn't really seem "exceptional". I come from a place where using exceptions for control flow is a big pessimization (C++) and I'm not sure how much of a concern it is in a language like python where performance is not a priority of the language.

So depending on how strongly you feel about keeping exceptions for exceptional situations, please consider refactoring.

Reply:
It's horrifically hacked. #6579 has a great example of how we might be able to update this code to use contextmanagers instead of these ridiculous implicit control flow, especially via exceptions.

Your comment is exactly on point, except for one thing, which is that we do expect this to be a truly *exceptional* case, in that it will only ever be thrown once in the lifetime of a pantsd client.

But @blorente described a solution which lets us avoid this chunk and segment entirely!! So it is moot, for now.....